asmcomp/amd64/emit.mlp

```diff
@@ -728,6 +728,19 @@ let emit_instr env fallthrough i =
       I.bswap (res i 0)
   | Lop(Ispecific(Ibswap _)) ->
       assert false
+  | Lop(Ispecific(Iclz 64)) ->
+      (* zero check not needed because the input is a tagged integer *)
+      if (arg i 0) <> (res i 0) then
+        I.xor (res32 i 0) (res32 i 0); (* avoid partial register stall *)
+      I.bsr (arg i 0) (res i 0);
+      I.xor (Imm 63L) (res i 0); (* convert BSR to CLZ *)
+      (* tag the result *)
```

Comment:
Tagging can be dealt with by `cmmgen`, if you declare the external like this:

```
external clz :  int -> (int[@untagged])  = "caml_int_clz_bytecode" "caml_int_clz" [@@noalloc]
```
and the C stubs would be like this
```
intnat caml_int_clz(value v)
{
  int count = 0;
  uintnat bit = (uintnat) 1 << (8 * sizeof(uintnat) - 1);
  uintnat u = v;
  /* The loop stops at the tagging bit. */
  while (!(u & bit))
    {
      ++count;
      bit >>= 1;
    }
  return Val_int(count);
  return count;
}

CAMLprim value caml_int_clz_bytecode(value v1)
{
  return Val_long(caml_int_clz(v1));
}
```

It may be worth having a faster C stub (using one of the methods in Hacker's Delight chap 5.3 or C compiler builtins as in [ocaml_intrinsics](https://github.com/janestreet/ocaml_intrinsics/blob/master/src/int_stubs.c) library).

Reply:
Interesting. I looked for something like this and couldn't find it. I'll make the change.

I'll check with the GCC developers if there is a code pattern that is recognized as `__builtin_clzl`, similar to what GCC does for population count. If we use that, we don't have to put intrinsics into `ints.c`.